### ISSUE TITLE
Newline Handling

### ISSUE THREAD

**Issue by indolering** [sunfishcode/basic-text#1]
This includes an elaborate rationale [primarily for historical reasons](https://fs.blog/2020/03/chestertons-fence/).  Feel free to jump to the TL;DR 😸.

The Unicode [newline technote](https://www.unicode.org/standard/reports/tr13/tr13-5.html#Recommendations) linked in the background document [was deprecated](https://www.unicode.org/standard/reports/tr13/) and folded into the standard some time ago (direct link: [latest](http://www.unicode.org/versions/latest/ch05.pdf#G10213), [Unicode 13](http://www.unicode.org/versions/Unicode13.0.0/ch05.pdf#G10213)).  I've copied the relevent sections here.

Note that the standard is a bit tricky to read, it defines New Line Function as `NLF ===  {CR, LF, CRLF, NEL}` which explicitly excludes the Unicode Line Function (NF) and Paragraph Separator (PS) and Form Feed (FF - telex lingo for _page separator_).

>The Unicode Standard defines two unambiguous separator characters: U+2029 paragraph separator (PS) and U+2028 line separator (LS). In Unicode text, the PS and LS characters should be used **wherever the desired function is unambiguous**. Otherwise, the following recommendations specify how to cope with an NLF when converting from other character sets to Unicode, when interpreting characters in text, and when converting from Unicode to other character sets.
>
>Note that **even if an implementer knows which characters represent NLF on a particular platform, CR, LF, CRLF, and NEL should be treated the same on input and in interpretation.** Only on output is it necessary to distinguish between them.
>
> ### Converting from Other Character Code Sets
> #### R1 If the exact usage of any NLF is known, convert it to LS or PS.
> #### R1a If the exact usage of any NLF is unknown, remap it to the platform NLF. 
> ### Interpreting Characters in Text
> _R2 - R2b: not relevant_
> #### R2c In parsing, choose the safest interpretation. _(relevant?)_
> For example, in recommendation R2c an implementer dealing with sentence break heuristics would reason in the following way that it is safer to interpret any NLF as LS:
> * Suppose an NLF were interpreted as LS, when it was meant to be PS. Because
most paragraphs are terminated with punctuation anyway, this would cause
misidentification of sentence boundaries in only a few cases.
> * Suppose an NLF were interpreted as PS, when it was meant to be LS. In this case, line breaks would cause sentence breaks, which would result in significant problems with the sentence break heuristics.
> ### Converting to Other Character Code Sets
> ####  R3 If the intended target is known, map NLF, LS, and PS depending on the target conventions. 
> For example, when mapping to Microsoft Word’s internal conventions for documents, LS would be mapped to VT, and PS and any NLF would be mapped to CRLF.
> #### R3a If the intended target is unknown, map NLF, LS, and PS to the platform newline convention (CR, LF, CRLF, or NEL).
> In Java, for example, this is done by mapping to a string nlf, defined as follows:
> String nlf = System.getProperty("line.separator");
> ### Input and Output
> #### R4 A readline function should stop at NLF, LS, FF, or PS. In the typical implementation, it does not include the NLF, LS, PS, or FF that caused it to stop.
> Because the separator is lost, the use of such a readline function is limited to text processing, where there is no difference among the types of separators.
> #### R4a A writeline (or newline) function should convert NLF, LS, and PS according to the recommendations R3 and R3a.
> In C, `gets` is defined to terminate at a newline and replaces the newline with '\0', while `fgets` is defined to terminate at a newline and includes the newline in the array into which it copies the data. C implementations interpret '\n' either as LF or as the underlying platform newline NLF, depending on where it occurs.  EBCDIC C compilers substitute the relevant codes, based on the EBCDIC execution set.
> ### Page Separator
> FF is commonly used as a page separator, and it should be interpreted that way in text. When displaying on the screen, it causes the text after the separator to be forced to the next page. It is interpreted in the same way as the LS for line breaking, in parsing, or in input segmentation such as readline. FF does not interrupt a paragraph, as paragraphs can and do span page boundaries.

My guess (based on my post HS graphic design job) is that Unicode's `U+2029` Paragraph Separator (PS) and `U+2028` Line Separator (LS) are **formatting controls**: most lay people insert two line separators after a paragraph to introduce whitespace.  This makes it impossible to (for example) indent the beginning sentence or reduce the distance between paragraphs to half the line height.  Dealing with formatting at the level of text encoding is a bit like using ACK for network signaling.  

R1a of _Converting from Other Character Code Sets_ states:
>If the exact usage of any NLF is unknown, remap it to the platform NLF.

**So not remapping LS, PS, or FF to NLF on input conforms with the standard.**

R3 and R3a of _Converting to Other Character Code Sets_ basically requires any mapping between NLF, LS, and PS to be configurable.

R4 of _Input and Output_ is trickier, as it entails processing NLF, LS, PS, or FF in a way that would be visible to the underlying program.  I have no informed opinion here, but I will note that pattern matching outside of ASCII can lead to performance issues.

Unicode [Regular Expressions](https://unicode.org/reports/tr18/#Line_Boundaries) and the Unicode [Line Breaking Algorithm](https://unicode.org/reports/tr14/#BK) appear to perform redundant checks for CR, LF, CRLF, NEL, LS, PS, and FF (so I don't think it matters how we handle them here.

TL;DR
=====
Unless there are compatibility or i18n issues that I am unaware of, defining WASI's New Line Function (NLF) as `U+000A` and remapping only `{CR, CRLF, NEL}` to `U+000A` (and **not** PS, LS, or FF) would follow the Unicode recommendation.  Compatibility concerns can be addressed through configuration.

**Comment by sunfishcode** [comment]
Thanks for filing this! I think everything here matches what's in the docs and code, except for the handling of U+85 (NEL). I was originally hoping we could avoid implementing it, but in addition to Unicode here, it's also in [XML 1.1](https://www.w3.org/TR/2006/REC-xml11-20060816/#sec-line-ends), so we should probably add compatible handling.

**Comment by sunfishcode** [comment]
Actually, the current rule is to translate U+85 to U+20, which ought to work in practice and may be sufficient. We may relax that if the need arises, but for now may be good enough.

**Comment by indolering** [comment]
> Actually, the current rule is to translate U+85 to U+20, which ought to work in practice and may be sufficient. We may relax that if the need arises, but for now may be good enough.

That is what Unicode recommends in R1a: "If the exact usage of any NLF is unknown, remap it to the platform NLF."
